src/dev/flang/ast/AbstractType.java

```diff
@@ -1562,6 +1564,64 @@ static boolean checkActualTypePars(SourcePosition pos, AbstractFeature called, L
     return result;
   }
 
+
+  /**
+   * collect all the directly and indirectly - via generics -
+   * used types in this type.
+   * example: Cons i32 (list i32) => { Cons, i32, list }
+   *
+   * @return
+   */
+  private Set<AbstractType> usedTypes()
+  {
+    var result = new TreeSet<AbstractType>();
+    usedTypes(result);
+    return result;
+  }
```

Comment:
generally, I prefer traversing on demand. There is a method `applyToGenericsAndOuter` that visits all used types (which is maybe too much), but you could add a variant `applyToXYZ`.
Here, I think all you need to check is the direct type parameters (see comments below), which are directly available with `generics()´.

Reply:
I now added an abstract method `abstract void usedFeatures(Set<AbstractFeature> s)` to `AbstractType` which collects all the features the type uses/needs into a set. These can then be checked for sufficient visibility.